plugins/inputs/socketstat/socketstat.go

```diff
@@ -0,0 +1,219 @@
+package socketstat
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+// Socketstat is a telegraf plugin to gather indicators from established connections, using iproute2's  `ss` command.
+type Socketstat struct {
+	BeginsWithBlank *regexp.Regexp
+	cmdName         string
+	Lister          socketLister
+	Log             telegraf.Logger
+	Measurement     string
+	SocketProto     []string
+	Timeout         config.Duration
+	ValidValues     *regexp.Regexp
+}
+
+type socketLister func(cmdName string, proto string, timeout config.Duration) (*bytes.Buffer, error)
+
+// Description returns a short description of the plugin
+func (ss *Socketstat) Description() string {
+	return "Gather indicators from established connections, using iproute2's `ss` command."
+}
+
+// SampleConfig returns sample configuration options
+func (ss *Socketstat) SampleConfig() string {
+	return `
+  ## ss can display information about tcp, udp, raw, unix, packet, dccp and sctp sockets
+  ## Specify here the types you want to gather
+  # socket_proto = [ "tcp", "udp" ]
+  ## The default timeout of 1s for ss execution can be overridden here:
+  # timeout = "1s"
+`
+}
+
+// Gather gathers indicators from established connections
+func (ss *Socketstat) Gather(acc telegraf.Accumulator) error {
+	// best effort : we continue through the protocols even if an error is encountered,
+	// but we keep track of the last error.
+	for _, proto := range ss.SocketProto {
+		out, e := ss.Lister(ss.cmdName, proto, ss.Timeout)
+		if e != nil {
+			acc.AddError(e)
+			continue
+		}
+		e = ss.parseAndGather(out, proto, acc)
+		if e != nil {
+			acc.AddError(e)
+			continue
+		}
+	}
+	return nil
+}
+
+func socketList(cmdName string, proto string, timeout config.Duration) (*bytes.Buffer, error) {
+	// Run ss for the given protocol, return the output as bytes.Buffer
+	args := []string{"-in", "--" + proto}
+	cmd := exec.Command(cmdName, args...)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := internal.RunTimeout(cmd, time.Duration(timeout))
+	if err != nil {
+		return &out, fmt.Errorf("error running ss -in --%s: ", err)
+	}
+	return &out, nil
+}
+
+func (ss *Socketstat) parseAndGather(data *bytes.Buffer, proto string, acc telegraf.Accumulator) error {
```

Comment:
Just a question here. AFAICS your parsing is that complicated due to the output format of `ss`. But you can influence this format. Did you try with `ss --no-header --oneline  ...` which provides the metrics with one-line per connection and without header?

Reply:
That's a good idea ! Unfortunately the --oneline option isn't present in 5.5.0 (e.g. Ubuntu 18.04 LTS), which was the version I was using when originally writing this plugin.
It's too early to deprecate such versions, so imho it's better to keep the code as is for now.

Reply:
TBH I have mixed feelings here. I guess the code is so much simpler with the new option and if we now add the more complex code, we will never simplify it again. So what about make the version `>=5.5.0` a requirement and add support for older versions as an external plugin for the meantime?

Reply:
In my use case, which is managing a few thousand servers of various OS versions, more than half of the machines would be excluded. So I have a strong preference to keep it this way for now (and for a few more years, until versions without --oneline are EOL for most distributions, e.g. Ubuntu, Debian, RH, etc.).